{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from __future__ import print_function\n",
    "import sys\n",
    "sys.path.append('./')\n",
    "import pandas as pd\n",
    "import subprocess\n",
    "import re\n",
    "import numpy as np\n",
    "import kang\n",
    "import pysam\n",
    "from tqdm import tqdm"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "ename": "ValueError",
     "evalue": "invalid literal for int() with base 10: '/root/.local/share/jupyter/runtime/kernel-8f59b421-a745-4cb1-a324-6ad6bc11eb3b.json'",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mValueError\u001b[0m                                Traceback (most recent call last)",
      "\u001b[1;32m<ipython-input-4-2cc008ebd5f1>\u001b[0m in \u001b[0;36m<module>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     15\u001b[0m \u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     16\u001b[0m \u001b[0mstrand\u001b[0m                  \u001b[1;33m=\u001b[0m \u001b[1;34m'+'\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[1;32m---> 17\u001b[1;33m \u001b[0mchromosome\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mleft\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mright\u001b[0m \u001b[1;33m=\u001b[0m \u001b[0msys\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0margv\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msys\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0margv\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m2\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mint\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0msys\u001b[0m\u001b[1;33m.\u001b[0m\u001b[0margv\u001b[0m\u001b[1;33m[\u001b[0m\u001b[1;36m3\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0m\u001b[0;32m     18\u001b[0m \u001b[0mgenename\u001b[0m                \u001b[1;33m=\u001b[0m \u001b[1;34m'%s_%d_%d'\u001b[0m\u001b[1;33m%\u001b[0m\u001b[1;33m(\u001b[0m\u001b[0mchromosome\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mleft\u001b[0m\u001b[1;33m,\u001b[0m \u001b[0mright\u001b[0m\u001b[1;33m)\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n\u001b[0;32m     19\u001b[0m \u001b[0mgeneseq\u001b[0m                 \u001b[1;33m=\u001b[0m \u001b[0mdic_fa\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mchromosome\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m[\u001b[0m\u001b[0mleft\u001b[0m\u001b[1;33m-\u001b[0m\u001b[1;36m1\u001b[0m\u001b[1;33m:\u001b[0m\u001b[0mright\u001b[0m\u001b[1;33m]\u001b[0m\u001b[1;33m\u001b[0m\u001b[0m\n",
      "\u001b[1;31mValueError\u001b[0m: invalid literal for int() with base 10: '/root/.local/share/jupyter/runtime/kernel-8f59b421-a745-4cb1-a324-6ad6bc11eb3b.json'"
     ]
    }
   ],
   "source": [
    "#+ preparing references\n",
    "df_gff_index = pd.read_pickle('./Athaliana_167_TAIR10.gene.gff3.pandas.df.pk')\n",
    "file_fa      = '/ref/analysis/DroughtNet/References/Creinhardtii/Creinhardtii_281_v5.0.fa'\n",
    "dic_fa       = kang.Fasta2dic(file_fa)\n",
    "df_gff_ix    = df_gff_index.reset_index().set_index(['genename','longest',2])\n",
    "df_gff_ix.sortlevel(inplace=True)\n",
    "\n",
    "genelist     = list(set(df_gff_ix.index.get_level_values('genename')))\n",
    "#- preparing references done\n",
    "\n",
    "\n",
    "#+ global vars.\n",
    "bamlist                 = ['/ref/analysis/Cre/braker/braker.try5_mario/intron3000.merge.sorted.bam',\\\n",
    "                           '/ref/analysis/stringtie2genemodel.results/Cre/30kintron/fixed/merged/all.merged.bam',\\\n",
    "                           '/ref/analysis/Cre/tophat/stranded_nitrogen_sulfur/hisat2/nitdef_0hr/SRR1521680/SRR1521680.sorted.bam',\\\n",
    "                           '/ref/analysis/Cre/tophat/stranded_nitrogen_sulfur/hisat2/nitdef_0hr/SRR1521685/SRR1521685.sorted.bam',\\\n",
    "                           '/ref/analysis/Cre/tophat/stranded_nitrogen_sulfur/unmapped_all/polya.reads.merged.fq.sorted.bam']\n",
    "\n",
    "\n",
    "strand                  = '+'\n",
    "\n",
    "chromosome, left, right = sys.argv[1], int(sys.argv[2]), int(sys.argv[3])\n",
    "\n",
    "genename                = '%s_%d_%d'%(chromosome, left, right)\n",
    "geneseq                 = dic_fa[chromosome][left-1:right]\n",
    "outfilename             = '%s_%d_%d.bamvisgene.svg'%(chromosome, left, right)\n",
    "Outfile                 = open(outfilename,'w')\n",
    "#- global vars. done\n",
    "\n",
    "#+ alignment vars.\n",
    "each_height   = 5\n",
    "each_space    = 2\n",
    "box_height    = 3\n",
    "bridge_height = 2\n",
    "genestart     = 3\n",
    "total_canvas_rows = 10 # 2 for gene model\n",
    "canvas_width      = 1000\n",
    "canvas_height     = (total_canvas_rows + 10)*(each_height+each_space)*len(bamlist) + 1000 # 1000 : just more blank\n",
    "real_width        = right-left+1+100+100\n",
    "#- alignments vars. done\n",
    "\n",
    "#+ svg configure\n",
    "rect     = '<rect x=\"%d\" y=\"%d\" width=\"%d\" height=\"%d\" style=\"fill:%s;stroke:%s;stroke-width:1;fill-opacity:1;stroke-opacity:0.3\" />'\n",
    "svg_line = '<line x1=\"%d\" y1=\"%d\" x2=\"%d\" y2=\"%d\" stroke=\"%s\" stroke-width=\"%d\" />'\n",
    "text     = '<text x=\"%d\"  y=\"%d\" style=\"font-family:Arial;font-size:%d;stroke:black;fill:black;\">%s</text>'\n",
    "#- svg configure done\n",
    "\n",
    "#+ color conf : haruhi theme\n",
    "c_utr     = '#f50002'\n",
    "c_cds     = '#53a1b5'\n",
    "c_fstrand = '#5d3c2d'\n",
    "c_sstrand = '#f5b024'\n",
    "#- color conf done\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def init_svg():\n",
    "    print('''<svg height=\"%d\" width=\"%d\">'''%(canvas_height,canvas_width),file=Outfile)\n",
    "def end_svg():\n",
    "    print('''</svg>\n",
    "    ''',file=Outfile)\n",
    "\n",
    "    Outfile.close()\n",
    "def get_ratio(x):\n",
    "    return int(canvas_width * float(x)/float(real_width))\n",
    "def get_ratio_f(x):\n",
    "    return (canvas_width * float(x)/float(real_width))\n",
    "def draw_words(y,strWord):\n",
    "    print(text%(15,y+8,10,strWord),file=Outfile)\n",
    "    return y+8+each_space+5\n",
    "def draw_gene(start_height,left,right,genename):\n",
    "    df = df_gff_ix.loc[(genename,'1')]\n",
    "    df = df.reset_index()\n",
    "    df = df.sort(3)\n",
    "\n",
    "    fill_CDS   = c_cds\n",
    "    strock_CDS = 'black'\n",
    "\n",
    "    fill_UTR   = c_utr\n",
    "    strock_UTR = 'black'\n",
    "\n",
    "    for i in df.index:\n",
    "        if df.loc[i][2] == 'mRNA':\n",
    "            strand  = df.loc[i][6]\n",
    "            mx1     = get_ratio(df.loc[i][3] - left + 100)\n",
    "            my1     = start_height + int(each_height/2) + genestart + 1\n",
    "            mx2     = get_ratio(df.loc[i][4] - left + 100)\n",
    "            my2     = start_height + int(each_height/2) + genestart + 1\n",
    "            print(svg_line%(mx1,my1,mx2,my2,'black',1),file=Outfile)\n",
    "    for i in df.index:\n",
    "        if 'UTR' in df.loc[i][2]:\n",
    "            ux1 = get_ratio(df.loc[i][3] - left + 100)\n",
    "            uy1 = start_height + genestart + 1\n",
    "            uw  = get_ratio(df.loc[i][4] - df.loc[i][3])\n",
    "            uh  = each_height\n",
    "            print(rect%(ux1,uy1,uw,uh,fill_UTR,strock_UTR),file=Outfile)\n",
    "        elif 'CDS' in df.loc[i][2]:\n",
    "            cx1 = get_ratio(df.loc[i][3] - left + 100)\n",
    "            cy1 = start_height + genestart + 1\n",
    "            cw  = get_ratio(df.loc[i][4] - df.loc[i][3])\n",
    "            ch  = each_height\n",
    "            print(rect%(cx1,cy1,cw,ch,fill_CDS,strock_CDS),file=Outfile)\n",
    "    print(text%(mx1,genestart+each_height*4,10,'%s (%s) %s:%d-%d'%(genename,strand,chromosome,left,right)),file=Outfile)\n",
    "    return start_height + genestart+each_height*4\n",
    "def match_count(a,b):\n",
    "        count = 0\n",
    "        if len(a)!=len(b):\n",
    "            return None\n",
    "        else:\n",
    "            for n,i in enumerate(a):\n",
    "                if a[n] == b[n]:\n",
    "                    count += 1\n",
    "                if a[n] != b[n]:\n",
    "                    break\n",
    "        return count\n",
    "def match_list(primer):\n",
    "    i = 0\n",
    "    j = 0\n",
    "    match_list = []\n",
    "    primer_seq = primer\n",
    "    while j < len(geneseq):\n",
    "        i = 0\n",
    "        inseq = geneseq[j:]\n",
    "        mc_list = []\n",
    "        while i != len(inseq):\n",
    "            mc = match_count(inseq[i:i+len(primer_seq)],primer_seq)\n",
    "            mc_list.append([i,mc])\n",
    "            i += 1\n",
    "        loc,mc = max(mc_list,key=lambda x:x[1])\n",
    "        match_list.append([j+loc,j+loc+mc])\n",
    "        j = j + loc + mc\n",
    "\n",
    "        primer_seq = primer_seq[mc:]\n",
    "        if primer_seq == '':\n",
    "            break\n",
    "    return match_list\n",
    "def draw_primer(start_height,primerlist,strand):\n",
    "    for strF,strR in primerlist:\n",
    "        if strand == '+':\n",
    "            strR = kang.rev_comp(strR)\n",
    "        else:\n",
    "            strF = kang.rev_comp(strF)\n",
    "        x_list = []\n",
    "        for x1,x2 in match_list(strF):\n",
    "            x_list.append(x1)\n",
    "            x_list.append(x2)\n",
    "            y1 = start_height + int(each_height/2)\n",
    "            y2 = start_height + int(each_height/2)\n",
    "            print(svg_line%(get_ratio(x1+100),y1,get_ratio(x2+100),y2,'blue',4),file=Outfile)\n",
    "        x_list.sort()\n",
    "        print(svg_line%(get_ratio(x_list[0]+100),y1,get_ratio(x_list[-1]+100),y2,'blue',1),file=Outfile)\n",
    "        x_list_2 = []\n",
    "        for x3,x4 in match_list(strR):\n",
    "            x_list_2.append(x3)\n",
    "            x_list_2.append(x4)\n",
    "            y1 = start_height + int(each_height/2)\n",
    "            y2 = start_height + int(each_height/2)\n",
    "            print(svg_line%(get_ratio(x3+100),y1,get_ratio(x4+100),y2,'red',4),file=Outfile)\n",
    "        x_list_2.sort()\n",
    "        print(svg_line%(get_ratio(x_list_2[0]+100),y1,get_ratio(x_list_2[-1]+100),y2,'red',1),file=Outfile)\n",
    "        if strand == '+':\n",
    "            print(svg_line%(get_ratio(x_list[-1]+100),y1,get_ratio(x_list_2[0]+100),y2,'gray',1),file=Outfile)\n",
    "        else: print(svg_line%(get_ratio(x_list_2[-1]+100),y1,get_ratio(x_list[0]+100),y2,'gray',1),file=Outfile)\n",
    "    return y2 + each_height\n",
    "def cigar_parse(cigar):\n",
    "    match = re.findall(r'(\\d+)(\\w)', cigar)\n",
    "    return match\n",
    "\n",
    "\n",
    "# pysam is problematic because it cannot handle efficiently the mate read information.\n",
    "# maybe it is better to have table in advance using pysam\n",
    "\n",
    "def draw_alignment(start_height,chromosome,left,right,file_bam,total_canvas_rows=5):\n",
    "    real_width        = right-left+1+100+100\n",
    "    gene_space        = np.zeros([total_canvas_rows,real_width])\n",
    "    samfile           = pysam.Samfile( file_bam, \"rb\" )\n",
    "    it                = samfile.fetch(chromosome,int(left),int(right))\n",
    "    dicFID2row = {} # fragment id : line.query_name\n",
    "    dicFID2count = {}\n",
    "    dicFID2line = {}\n",
    "    for line in tqdm(it):\n",
    "        if line.is_paired == True and line.is_proper_pair == False:\n",
    "            continue\n",
    "        if line.is_duplicate   == True:\n",
    "            continue\n",
    "        if line.is_qcfail      == True:\n",
    "            continue\n",
    "        if line.is_secondary   == True:\n",
    "            continue\n",
    "        try:\n",
    "            dicFID2line[line.query_name].append(line)\n",
    "        except KeyError:\n",
    "            dicFID2line[line.query_name] = [line]\n",
    "    dicFID2line_list = dicFID2line.keys()\n",
    "    dicFID2line_list.sort(key=lambda x : int(dicFID2line[x][0].reference_start))\n",
    "    for fragmentid in tqdm(dicFID2line_list):\n",
    "        lines = dicFID2line[fragmentid]\n",
    "        lines.sort(key=lambda x : int(x.reference_start))\n",
    "        line  = lines[0]\n",
    "        fragstart      = line.reference_start-left+1+100\n",
    "        if fragstart < 0 :\n",
    "            continue\n",
    "        cigar          = line.cigarstring\n",
    "        try:\n",
    "            cigarM       = cigar_parse(cigar)\n",
    "        except TypeError:\n",
    "            continue\n",
    "        cigarstrings  = [x[1] for x in cigarM]\n",
    "        cigarvalues   = [x[0] for x in cigarM]\n",
    "\n",
    "        if line.is_paired == True:\n",
    "            if line.is_proper_pair == False:\n",
    "                continue\n",
    "            if len(lines) < 2:\n",
    "                continue\n",
    "            try:\n",
    "                fragmentsize   = line.mpos+lines[1].reference_length - line.reference_start # reference length of fragment (intron added)\n",
    "            except TypeError:\n",
    "                continue\n",
    "            bridge_length  = line.mpos - line.reference_start - line.reference_length\n",
    "            if fragmentsize > 0 and bridge_length < 10 :\n",
    "                continue\n",
    "        else:\n",
    "            fragmentsize   = line.reference_length\n",
    "\n",
    "        if abs(fragmentsize) == 0 :\n",
    "            continue\n",
    "        if fragmentsize < 0 :\n",
    "            print(fragmentid,len(lines),cigar,line.mpos,line.reference_start,line.reference_length)\n",
    "            print(1111)\n",
    "            exit()\n",
    "\n",
    "        #+ row selection\n",
    "        try:\n",
    "            srow = dicFID2row[fragmentid]\n",
    "        except KeyError:\n",
    "            srow = None\n",
    "            for nrow, row in enumerate(gene_space):\n",
    "                if 1:\n",
    "                    if max(row[fragstart:fragstart+fragmentsize]) > 0 :\n",
    "                        continue\n",
    "                    else:\n",
    "                        srow = nrow\n",
    "                        gene_space[srow,fragstart:fragstart+fragmentsize] += 1\n",
    "                        break\n",
    "            dicFID2row[fragmentid] = srow\n",
    "\n",
    "        if srow == None:\n",
    "            continue\n",
    "        #- row selection\n",
    "\n",
    "\n",
    "        if line.is_paired == True:\n",
    "            if fragstart < 0:\n",
    "                continue\n",
    "            #keepdoing      = 1\n",
    "            #for i in np.arange(fragstart,fragstart+fragmentsize):\n",
    "            #    try:\n",
    "            #        flatten = np.sum(gene_space[:,i])\n",
    "            #    except IndexError:\n",
    "            #        keepdoing = 0\n",
    "            #    if flatten == total_canvas_rows:\n",
    "            #        keepdoing = 0\n",
    "            #if keepdoing == 0 :\n",
    "            #    continue\n",
    "            if line.is_read1 == True and line.is_reverse == False:\n",
    "                blockcolor = c_sstrand\n",
    "            elif line.is_read2 == True and line.is_reverse == True:\n",
    "                blockcolor = c_sstrand\n",
    "            elif line.is_read1 == True and line.is_reverse == True:\n",
    "                blockcolor = c_fstrand\n",
    "            elif line.is_read2 == True and line.is_reverse == False:\n",
    "                blockcolor = c_fstrand\n",
    "\n",
    "            #+ draw bridge for PE read 1\n",
    "            if 1:\n",
    "                coverleng = 0\n",
    "                for j, cigarstring in enumerate(cigarstrings):\n",
    "                    if cigarstring == 'M' or cigarstring == 'N' or cigarstring == 'I':\n",
    "                        coverleng += int(cigarvalues[j])\n",
    "                if 1:\n",
    "                    x1 = get_ratio(fragstart + coverleng)\n",
    "                    y1 = start_height + srow * (each_height + each_space) + int(each_height/2)\n",
    "                    y2 = start_height + srow * (each_height + each_space) + int(each_height/2) + int(each_height/2)\n",
    "                    x3 = get_ratio(line.mpos - left+1 + 100)\n",
    "                    x2 = (x3 + x1)/2\n",
    "                    print(svg_line%(x1,y1,x2,y2,'black',1),file=Outfile)\n",
    "                    print(svg_line%(x2,y2,x3,y1,'black',1),file=Outfile)\n",
    "                    #print(text%(x1,y1,5,fragmentid),file=Outfile)\n",
    "            #- draw bridge for PE\n",
    "\n",
    "        ##################################### cigar check start\n",
    "\n",
    "            #+ draw match and gap\n",
    "            for eline in lines:\n",
    "                w = 0\n",
    "                fragstart = eline.reference_start - left + 1 + 100\n",
    "                cigarstrings = eline.cigarstring\n",
    "                cigar        = eline.cigarstring\n",
    "                try:\n",
    "                    cigarM       = cigar_parse(cigar)\n",
    "                except TypeError:\n",
    "                    continue\n",
    "                cigarstrings  = [x[1] for x in cigarM]\n",
    "                cigarvalues   = [x[0] for x in cigarM]\n",
    "\n",
    "                for j, cigarstring in enumerate(cigarstrings):\n",
    "                    if cigarstring == 'M':\n",
    "                        x = get_ratio(w + fragstart) # get ratio have gap with actual value\n",
    "                        if x < 0:\n",
    "                            continue\n",
    "                        y = start_height + srow * (each_height + each_space)\n",
    "                        h = int(each_height)\n",
    "                        ew = int(cigarvalues[j]) # each width\n",
    "                        w += ew\n",
    "                        print(rect%(x,y,get_ratio(ew),h,blockcolor,'black'),file=Outfile)\n",
    "                    elif cigarstring == 'N' or cigarstring == 'I':\n",
    "                        x1 = get_ratio(w + fragstart)\n",
    "                        if x1 < 0:\n",
    "                            continue\n",
    "                        y1 = start_height + srow * (each_height + each_space) + int(each_height/2)\n",
    "                        ew = int(cigarvalues[j]) # each width\n",
    "                        w += ew\n",
    "                        x2 = get_ratio(w + fragstart)\n",
    "                        print(svg_line%(x1,y1,x2,y1,'pink',1),file=Outfile)\n",
    "            #- draw match and gap\n",
    "        else:\n",
    "            fragstart      = line.reference_start-left+1+100\n",
    "            if fragstart < 0:\n",
    "                continue\n",
    "\n",
    "            if line.is_reverse == True:\n",
    "                    blockcolor = c_sstrand\n",
    "            else:\n",
    "                    blockcolor = c_fstrand\n",
    "            w = 0\n",
    "            for j, cigarstring in enumerate(cigarstrings):\n",
    "                if cigarstring == 'M':\n",
    "                    x = get_ratio(w + fragstart)\n",
    "                    if x < 0:\n",
    "                        continue\n",
    "                    y = start_height + srow * (each_height + each_space)\n",
    "                    h = int(each_height)\n",
    "                    ew = int(cigarvalues[j]) # each width\n",
    "                    w += ew\n",
    "                    print(rect%(x,y,get_ratio(ew),h,blockcolor,'black'),file=Outfile)\n",
    "                elif cigarstring == 'N' or cigarstring == 'I':\n",
    "                    x1 = get_ratio(w + fragstart)\n",
    "                    if x1 < 0:\n",
    "                        continue\n",
    "                    y1 = start_height + srow * (each_height + each_space) + int(each_height/2)\n",
    "                    ew = int(cigarvalues[j]) # each width\n",
    "                    w += ew\n",
    "                    x2 = get_ratio(w + fragstart)\n",
    "                    print(svg_line%(x1,y1,x2,y1,'pink',1),file=Outfile)\n",
    "    return start_height+(each_height + each_space) * len(gene_space)\n",
    "\n",
    " \n",
    " "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "init_svg()\n",
    "#endp = draw_gene(5,left,right,genename)\n",
    "#endp = draw_primer(endp,primerlist,strand)\n",
    "endp = 5\n",
    "for bam in bamlist:\n",
    "    endp = draw_alignment(endp,chromosome,left,right,bam,20)\n",
    "    endp = draw_words(endp,bam.split('/')[-1])\n",
    "    #print(1)\n",
    "end_svg()\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
